### PR TITLE
Fixes Bug #63814: Allow for index creation with USING and WHERE keywords...

### PR DESCRIPTION
**Pull request by SPodjasek**
... supplied as add_index options

### CODE REVIEW

**Comment by SPodjasek** [comment]
Current PostgreSQL version (9.4) allows some other options to be set:

    CREATE [ UNIQUE ] INDEX [ CONCURRENTLY ] [ name ] ON table_name [ USING method ]
        ( { column_name | ( expression ) } [ COLLATE collation ] [ opclass ] [ ASC | DESC ] [ NULLS { FIRST | LAST } ] [, ...] )
        [ WITH ( storage_parameter = value [, ... ] ) ]
        [ TABLESPACE tablespace_name ]
        [ WHERE predicate ]

but nobody asked for them for now

File: lib/SQL/Translator/Producer/PostgreSQL.pm [hunk 1]
```diff
@@ -553,7 +570,10 @@ sub create_index
     }
     elsif ( $type eq NORMAL ) {
         $index_def =
-            'CREATE INDEX ' . $generator->quote($name) . ' on ' . $generator->quote($table_name) . ' ' . $field_names
+            'CREATE INDEX ' . $generator->quote($name) . ' on ' . $generator->quote($table_name) .
+            ($index_using ne '' ? ' ' . $index_using : '') .
```
Comment (ilmari) [inline lib/SQL/Translator/Producer/PostgreSQL.pm]: If you include the leading space when setting the value above, you can get rid of the ugly conditionals here.

**Comment by ilmari** [comment]
Not a prerequisite for merging this, but it would be nice to support this in `Parser::PostgreSQL` as well. The grammar already parses the clauses it just doesn't store them in the `Schema::Index` object

**Comment by SPodjasek** [comment]
Looking at `Parser::PostgreSQL` they already store index method from `USING` - but in key named `method`, not in `options` array

**Comment by ilmari** [comment]
It's storing it in the data structure it builds during parsing, but it's not propagating that in the `$table->add_index` call.

**Comment by ribasushi** [comment]
@SPodjasek @ilmari This is also likely related to https://rt.cpan.org/Ticket/Display.html?id=101704 and https://github.com/frioux/DBIx-Class-DeploymentHandler/issues/32#issue-55127029

**Comment by ilmari** [comment]
@ribasushi No, that's about specifying an explicit conversion when changing column types, it just happens to also use the `USING` keyword.

**Comment by SPodjasek** [comment]
My Parser::PostgreSQL changes lack support of function fields in indexes, but it is also an issue with Parser::DBI::PostgreSQL (as stated in line 144 there)

**Comment by ilmari** [comment]
Thank you, merged with minor tweaks as 5b36314 and 9c05d80.

**Pull request closed without merging**